Added CodeStyle option for 'using' placement

JoeRobich opened the pull request:
User is able to specify Preserve, Inside Namespaces, or Outside Namespaces. Default is Preserve.

Move Inside Namespace enabled:
![move usings inside namespace](https://user-images.githubusercontent.com/611219/47581627-c3a1eb80-d906-11e8-8d4f-5d004f80d683.gif)

Move Outside Namespace enabled:
![move usings outside namespace](https://user-images.githubusercontent.com/611219/47581638-cf8dad80-d906-11e8-915b-74e815f38c36.gif)

Move Usings Options:
![move usings code style options](https://user-images.githubusercontent.com/611219/47585734-023da300-d913-11e8-8636-43b7ea221a57.gif)

resolves #27179

/CC: @sharwell

CyrusNajmabadi commented:
> User is able to specify No Preference, Inside Namespaces, or Outside Namespaces. Default is No Preference.

We don't really ever have a 'no preference' option. There's always a preference, there is just a configured severity that determines if the user is pushed to fix things up or not.  Should we just have "Inside/Output" along with a severity of 'None'?

CyrusNajmabadi commented on src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj:
```diff
@@ -11,6 +11,7 @@
   <ItemGroup>
     <Compile Include="..\..\..\Features\Core\Portable\CodeFixes\PredefinedCodeFixProviderNames.cs" Link="PredefinedCodeFixProviderNames.cs" />
     <Compile Include="..\..\..\Workspaces\Core\Portable\Shared\Utilities\IProgressTracker.cs" Link="IProgressTracker.cs" />
+    <Compile Include="..\..\..\Workspaces\Core\Portable\CodeFixes\FixAllOccurrences\DocumentBasedFixAllProvider.cs" Link="DocumentBasedFixAllProvider.cs" />
```
why?

sharwell replied:
➡️ The original implementation was based on this approach. We already had the type in the code base so it was linked it to support the new code fix.

CyrusNajmabadi replied:
I'm not actually seeing FixAll tests, so i'm not sure how to evaluate this.  It's also unclear to me why this is in src/CodeStyle/Core since  the feature doesn't appear to be there.

JoeRobich replied:
`DocumentBasedFixAllProvider` was already in CodeStyle.Fixes. I moved it to Workspaces to make it accessible to the fixAll and other future fixAlls. I could have alternately refernced CodeStyle.Fixes from the CSharp.EditorFeatures project.

sharwell replied:
> CodeStyle.Fixes

This is not currently possible due to #30472

CyrusNajmabadi commented on src/EditorFeatures/CSharp/CSharpEditorResources.resx:
```diff
@@ -117,6 +117,18 @@
+  </data>
+  <data name="Move_misplaced_using_statements" xml:space="preserve">
+    <value>Move misplaced using statements</value>
+    <comment>{Locked="using"} "using" is a C# keyword and should not be localized.</comment>
```
what is the difference from "Move misplaced" and "Fix misplaced"?

Neme12 replied:
This should be using directives, not statements right?

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.IndentationSettings.cs:
```diff
@@ -0,0 +1,41 @@
+        {
+            public int IndentationSize { get; }
+            public int TabSize { get; }
+            public bool UseTabs { get; }
```
this seems wacky to pass along when we have an Options type that holds this sort of thing.

sharwell replied:
This was an artifact of the original implementation not having access to .editorconfig. Code using this type could be updated in Roslyn to use the .editorconfig values instead.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.SourceMap.cs:
```diff
@@ -0,0 +1,281 @@
+    internal partial class MisplacedUsingsCodeFixProvider
+    {
+        /// <summary>
+        /// Contains a map of the different regions of a source file.
```
what is this needed for?

sharwell replied:
The ordering mechanism allows using directives to be grouped into regions, within with sort order is preserved. Ordering within one group is not checked with respect to other groups.

CyrusNajmabadi replied:
i can't make heads or tails of htat :)

What does ordering/grouping/etc. have to do with moving usings from outside a namespace to inside (or vice versa)?

CyrusNajmabadi replied:
Gotcha.  That's currently definitely *not* the Roslyn way.  The Roslyn way is to have a general command called "organize usings" and individual operatoins internally which build that up.  i.e. sorting, grouping, removing unused.

My expectation would be that we would:
1. add the individual functionality to 'move' usings to the right location.
2. update 'organize usings' to respect that.  (note: this wasn't done in this PR, but it really need to be.
3. Have an analyzer that checks if your usings are *not organized* and offers to fully organize them.

'3' would effectively be equivalent to what you were saying about SyleCopAnalyzers.  Except it would be about all the using-organizaiton options, and not just aobut "are the usings in the right container".

@JoeRobich I would basically expect all this given that we already have the other style options, and we already have the "organize" command.  Right now, this PR feels disjointed because thing aren't fitting together in a common way.

CyrusNajmabadi commented:
Pausing my review.  can you give a medium level overview of the implementation of this feature?  There was a lot of unexpected complexity, and it would be good to have an understanding of what it is for, why existing facilities could not be used, and how ti all works.  Thanks!

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/CodeStyle/UsingPlacementPreference.cs:
```diff
@@ -0,0 +1,22 @@
+        ///<summary>
+        /// Prefer usings placed outside namespaces.
+        ///</summary>
+        OutsideNamespace
```
better names might be 'CompilationLevel' and 'NamespaceLevel'.  You may not have a namespace in your code, so saying 'OutputNamespace' doesn't really make a ton of sense to me.

CyrusNajmabadi replied:
With only two values, this could potentially also be a bool.  though i' mnot opposed to an enum.

sharwell replied:
I don't see the suggested alternatives as a clear improvement over the current names. The current names align with the documentation described by both StyleCop Classic and StyleCop Analyzers, so it seems consistent with what users would likely expect today.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/CSharpWorkspaceResources.resx:
```diff
@@ -168,4 +168,7 @@
     <value>Wrapping preferences</value>
   </data>
+  <data name="using_preferences" xml:space="preserve">
+    <value>using preferences</value>
```
no comment that 'using' should be locked?

Neme12 replied:
Should `using` be in single quotes?

JoeRobich replied:
@Neme12  The `var` in `var_preferences` isn't in quotes, so I think it is OK.

CyrusNajmabadi replied:
one niggle there is htat 'var preferences' isn't really an english phrase.  whereas 'using preferences' is a bit ambiguous.  Is it hte preferences around 'using directives'?  Or is it 'using your preferences'?  That said, this is minor enough that i think it's fine either way.

CyrusNajmabadi replied:
looking at this, i would put 'using' in quotes.

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs:
```diff
@@ -86,6 +86,8 @@ internal static class IDEDiagnosticIds
 
         public const string FormattingDiagnosticId = "IDE0054";
 
+        public const string MoveMisplacedUsingsDiagnosticId = "IDE0055";
```
will conflict with other PRs going in.

Neme12 replied:
Idea: If blank lines between these were removed, it would show up as an actual conflict so that people would be alerted of this and fix it.

Neme12 replied:
Actually it should be a conflict anyway, forget what I'm saying.

JoeRobich replied:
Is there an easy way to ask GitHub for all the open PRs that modify this file? =)

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            return s_noPreferenceOption;
+        }
+
+        private bool ShouldSurpressDiagnostic(CompilationUnitSyntax compilationUnit)
```
"Suppress" (two 'p's, no 'r').

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+        {
+            // Surpress if file contains a type declaration in the global namespace
+            // or if an attribute is used in the global namespace.
+            return compilationUnit.ChildNodes().Any(node =>
```
wouldn't a simpler way be to just check if it's not just usings and a single namespace?

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+                || node.IsKind(SyntaxKind.EnumDeclaration)
+                || node.IsKind(SyntaxKind.StructDeclaration)
+                || node.IsKind(SyntaxKind.DelegateDeclaration)
+                || node.IsKind(SyntaxKind.AttributeList));
```
top level might also include things like members.  instead of a blacklist, make a whitelist.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            var filePath = context.SemanticModel.SyntaxTree.FilePath;
+            var fullPath = string.IsNullOrEmpty(Path.GetDirectoryName(filePath))
+                ? Path.Combine(Environment.CurrentDirectory, filePath)
+                : filePath;
```
i don't understand this.  NO other features do this, so it feels very wrong to me for this feature to do this.

sharwell replied:
➡️ Requirement for injecting .editorconfig values into the new test framework, prior to full adoption of the upcoming compiler-layer APIs for .editorconfig. We tried hard to avoid this, but all options (aside from waiting for the new APIs) were ugly.

CyrusNajmabadi replied:
This should be encapsulated somewhere so that this can happen without each individual feature needing to have hardcoded in knowledge of this sort of hting.  That way the brittleness is isolated to a single location, isntead of virally spreading into every feature.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            // Find the correct key name from the .editorconfig storage location
+            var storageLocation = CSharpCodeStyleOptions.PreferredUsingPlacement.StorageLocations
+                .OfType<EditorConfigStorageLocation<CodeStyleOption<UsingPlacementPreference>>>()
+                .FirstOrDefault();
```
what is going on here?  In general, there's no way for you to know the implementation of StorageLocations.  This is taking a hard dependency on there being such a config location value here.

JoeRobich replied:
I was trying to avoid hardcoding the .editorconfig option key into the analyzer. Perhaps a better long term solution is pulling all the .editorconfig option keys into a file we could reference as constants.

CyrusNajmabadi replied:
I guess my point is this:  we already ahve the Options object.  So why are we peaking inside hte options object assumign it has a certain shape.  The Options object is already supposed to be the thing that gives you the values back.  So why is it not just doing all this work inside of itself?

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+                ReportDiagnostics(context, _insideDescriptor, compilationUnit.Usings, option);
+            }
+            else if (statement is NamespaceDeclarationSyntax namespaceDeclaration
+                && option.Value == UsingPlacementPreference.OutsideNamespace)
```
can you alight parts of a conditoin with each other.  also, the vast majority of IDE code puts boolean operators on previous line.  can you be consistent with that?

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/MisplacedUsings/MisplacedUsingsCodeFixProviderTests.cs:
```diff
@@ -0,0 +1,1295 @@
+    /// <summary>
+    /// Unit tests for the <see cref="MisplacedUsingsDiagnosticAnalyzer"/> and <see cref="MisplacedUsingsCodeFixProvider"/>.
+    /// </summary>
+    public class MisplacedUsingsCodeFixProviderTests
```
i'm confused why this isn't just using our normal code fix base class.

sharwell replied:
Most of the deviations on this front are the result of avoiding unnecessary deviations from the original implementation in the initial submission. There was no "normal code fix base class" in the original.

CyrusNajmabadi replied:
I'm ok with that as a first draft.  But it does'nt make sense to me in terms of merging in.  Will we just be producing more and more duplication of code that may already exist just because it's in StyleCop analyzers?  I def do not like that. This just means more code to maintain and fixup when we encounter bugs and whatnot.

sharwell replied:
Roslyn is one of the remaining outlies that does not use Microsoft.CodeAnalysis.Testing, the new common library for analyzer and code fix testing. The default configuration for that library includes a large number of much stronger assertions regarding "well-behaved" implementations, so the long term goal is migrating all current testing to be built on that library.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/MisplacedUsings/MisplacedUsingsCodeFixProviderTests.cs:
```diff
@@ -0,0 +1,1295 @@
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(6, 1),
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(7, 1),
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(9, 1),
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(10, 1),
```
1. why are we reporting so many misplaced diagnostics?  Users should really only see one diagnostic (either for the span of the first item, or perhaps the span of the entire set).
2. I don't like this test because it's no longer declarative.  i have no idea what `10,1` means, or how it relates to the code.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/MisplacedUsings/MisplacedUsingsCodeFixProviderTests.cs:
```diff
@@ -0,0 +1,1295 @@
+}
+";
+
+            DiagnosticResult[] expected =
```
var

sharwell replied:
➡️ Would result in a compile error here.

CyrusNajmabadi replied:
Why?

Neme12 replied:
because it's using an array initializer instead of `new[]`?

CyrusNajmabadi replied:
ah, my mind read it simply as "new [] {".  Can we switch to that as the vast majority of IDE code uses 'var' when possible.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/MisplacedUsings/MisplacedUsingsCodeFixProviderTests.cs:
```diff
@@ -0,0 +1,1295 @@
+    using static System.String;
+
+    using MyFunc = System.Func<int,bool>;
+    using SystemAction = System.Action;
```
why are we doing more than we said we would do.  i'm not a big fan of 'move usings' now applying other things.  We already have 'organize' usings, which shoudl be doing the whole "move, sort, add blank lines".  But specific actions like "remove usings" "sort usings" or "move usings" should just do those single things.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/MisplacedUsings/MisplacedUsingsCodeFixProviderTests.cs:
```diff
@@ -0,0 +1,1295 @@
+
+            var fixedTestCode = @"namespace TestNamespace
+{
+#pragma warning disable 1573 // Comment
```
why does the pragma move with the usings?

sharwell replied:
:arrow_right: It could be suppressing a diagnostic related to the using directive itself

CyrusNajmabadi replied:
wouldn't that still happen even if it didn't move?  what if its suppressing something related to the namespace it's moving into?

CyrusNajmabadi commented:
> The tests cover nearly all branches of the code fix; it turned out to be surprisingly complex for what seemed like a reasonably straightforward feature going in.

Hrmm... I just looked at all the tests.  And nothing about them is striking me as warranting the extra complexity i'm seeing here.  It may be the case that it's necessary due to internal code not being available for StyelCopAnalyzers to use.  And that would indeed be a pity if that's the case.  However, i'm not a fan at all of the idea of a large mount of duplication and/or unnecessary code flowing *into* roslyn if that's the case.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.IndentationHelper.cs:
```diff
@@ -0,0 +1,84 @@
+                    result = new string(' ', indentationCount);
+                }
+
+                return result;
```
this seems duplicated.  IDE already has code for this.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.IndentationSettings.cs:
```diff
@@ -0,0 +1,41 @@
+            {
+                this.TabSize = options.GetOption(FormattingOptions.TabSize, LanguageNames.CSharp);
+                this.IndentationSize = options.GetOption(FormattingOptions.IndentationSize, LanguageNames.CSharp);
+                this.UseTabs = options.GetOption(FormattingOptions.UseTabs, LanguageNames.CSharp);
```
why not just pass FormattingOptions along to the code that needs this?

JoeRobich replied:
Sorry I am not quite sure what is being suggested here.  Is the suggestion to pass these three values around instead of an instance of this class?

CyrusNajmabadi replied:
The suggestion is to just pass around the OptionSet .  It's already immutable and easy to pass along.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.SourceMap.cs:
```diff
@@ -0,0 +1,281 @@
+        /// <item><description>region directives</description></item>
+        /// </list>
+        /// </remarks>
+        private class SourceMap
```
this info feels like it could be very redundant with the info we compute in DirectiveSyntaxExtensions.

JoeRobich commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+
+            if (statement is CompilationUnitSyntax compilationUnit
+                && option.Value == UsingPlacementPreference.InsideNamespace
+                && !ShouldSurpressDiagnostic(compilationUnit))
```
Was trying to remove duplication of the code gets the placement preference since it is a majority of the code, but i see now it wasn't an issue. =) Thanks!

JoeRobich commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.cs:
```diff
@@ -0,0 +1,570 @@
+            if (CountNamespaces(compilationUnit.Members) > 1)
+            {
+                return;
+            }
```
The using placement goes against the code style preference, but the code fix doesn't currently do the work of determining which namespaces to move each `using` into.

CyrusNajmabadi replied:
enhhhhhhhhh.  i would think it would just be the outermost namespace.  do we have people with another coding pattern somewhere?

sharwell replied:
> enhhhhhhhhh

It's safe to assume that everything you hit like this was implemented as a direct result of a bug report filed to StyleCop Analyzers. This feature took more than a year to stabilize with some rather horrifying real-world edge cases.

JoeRobich commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.UsingsSorter.cs:
```diff
@@ -0,0 +1,611 @@
+                var root = GetNamespaceRoot(previousUsing.Name);
+
+                // We prime with the first using's Namespace root and the last using will always get a blank line.
+                for (var i = 1; i < usingsList.Count - 1; i++)
```
I should look at the last item as well.

sharwell commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.UsingsSorter.cs:
```diff
@@ -343,8 +343,8 @@ private void InsertBlankLinesBetweenSubGroups(List<UsingDirectiveSyntax> usingsL
-                // We prime with the first using's Namespace root and the last using will always get a blank line.
-                for (var i = 1; i < usingsList.Count - 1; i++)
+                // We have already processed the first using's Namespace root.
+                for (var i = 1; i < usingsList.Count; i++)
```
💭 Why did this change from `usingsList.Count - 1` to `usingsList.Count`?

JoeRobich replied:
I was initially wrong in thinking I didn't have to look at the last node. The blank line is applied to the previous node, so a gap could be required.

sharwell replied:
Can you make sure we have a test covering this case (i.e. would fail if this was changed back to `Count - 1`)?

sharwell commented on src/Workspaces/CSharp/Portable/CodeStyle/CSharpCodeStyleOptions.cs:
```diff
@@ -187,6 +187,19 @@ private static Option<T> CreateOption<T>(OptionGroup group, string name, T defau
+            defaultValue: NoPreferenceWithSilentEnforcement,
+            storageLocations: new OptionStorageLocation[]{
+                new EditorConfigStorageLocation<CodeStyleOption<UsingPlacementPreference>>(
+                    "csharp_preferred_using_placement",
```
💭 `csharp_using_directive_placement`?

sharwell commented on src/Workspaces/CSharp/Portable/CodeStyle/CSharpCodeStyleOptions_Parsing.cs:
```diff
@@ -49,5 +50,47 @@ private static string GetExpressionBodyPreferenceEditorConfigString(CodeStyleOpt
+                {
+                    switch (value)
+                    {
+                        case "no_preference":
```
❓ Why not `preserve`? The current name sounds like using directives are not allowed (which is actually a code style I've seen before).

sharwell commented on src/VisualStudio/CSharp/Impl/xlf/CSharpVSResources.ru.xlf:
```diff
@@ -42,6 +57,11 @@
         <note>'is null' is a C# string and should not be localized.</note>
       </trans-unit>
+      <trans-unit id="Preferred_usings_placement">
+        <source>Preferred usings placement</source>
```
💭 'Using directives placement'?

JoeRobich commented:
@CyrusNajmabadi & @sharwell  This PR is ready for another review. 

Notable changes since October:
- Replaced the StyleCop fixer (Sort, Group, and Move) to a fixer that just moves using directives
- Split the Analyzer into one for the CompilationUnit and one for Namespaces
- Updated the AddImportService to respect preferred placement

CyrusNajmabadi commented on eng/Versions.props:
```diff
@@ -25,7 +25,7 @@
     <RoslynDiagnosticsNugetPackageVersion>2.6.2-beta2</RoslynDiagnosticsNugetPackageVersion>
     <CodeStyleLayerCodeAnalysisVersion>2.8.2</CodeStyleLayerCodeAnalysisVersion>
-    <MicrosoftCodeAnalysisTestingVersion>1.0.0-beta1-63310-01</MicrosoftCodeAnalysisTestingVersion>
+    <MicrosoftCodeAnalysisTestingVersion>1.0.0-beta1-63812-02</MicrosoftCodeAnalysisTestingVersion>
```
why?

sharwell replied:
Newer versions of this library often come with usability enhancements for testing analyzers and code fixes.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.cs:
```diff
@@ -0,0 +1,405 @@
+            }
+        }
+
+        private class MoveMisplacedUsingsCodeAction : CodeAction
```
we already have a CodeAction.DocumentChangeAction that we derive from for these cases.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Diagnostics/CodingConventionAnalyzerHelpers.cs:
```diff
@@ -0,0 +1,58 @@
+            // Provide a rooted path when filePath is just the file's name.
+            var fullPath = string.IsNullOrEmpty(Path.GetDirectoryName(filePath))
+                ? Path.Combine(Environment.CurrentDirectory, filePath)
+                : filePath;
```
seeing Environment in a test is scary to me.

CyrusNajmabadi commented on src/Features/Core/Portable/AddImport/References/PackageReference.cs:
```diff
@@ -28,10 +29,10 @@ private partial class PackageReference : Reference
 
             public override async Task<AddImportFixData> TryGetFixDataAsync(
-                Document document, SyntaxNode node, bool placeSystemNamespaceFirst, CancellationToken cancellationToken)
+                Document document, SyntaxNode node, bool placeSystemNamespaceFirst, AddImportPlacement placement, CancellationToken cancellationToken)
```
consider an AddImportOptions struct that can pass along the placement and the PlaceSystemNamespaceFirst value.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/Formatting/StyleViewModel.cs:
```diff
@@ -1534,6 +1576,13 @@ internal StyleViewModel(OptionStore optionStore, IServiceProvider serviceProvide
+                new CodeStylePreference(CSharpVSResources.Preserve, isChecked: false),
+                new CodeStylePreference(CSharpVSResources.Inside_namespace, isChecked: false),
+                new CodeStylePreference(CSharpVSResources.Outside_namespace, isChecked: false),
+            };
```
having a 'preserve' option seems odd to me.  we don't generlaly have that.  our position up till now has been that we always have a preference, but we can have no-severity, meaning we won't tell you there's an issue, or otherwise try to get you to change it.

JoeRobich replied:
Sam brought up that when we update Organize Imports it would use the Preserve option to mean don't move imports. There is also the Add Import code fix, if we didn't have the Preserve option then the current location of using directives would get ignored in favor of the code style setting we defaulted to.

CyrusNajmabadi replied:
> here is also the Add Import code fix, if we didn't have the Preserve option then the current location of using directives would get ignored in favor of the code style setting we defaulted to.


Hrmm... for add-import, i think we should respect the file no matter what the option is.  Thoughts?

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/CodeStyle/CSharpCodeStyleOptions.cs:
```diff
@@ -227,6 +228,19 @@ private static Option<T> CreateOption<T>(OptionGroup group, string name, T defau
+            defaultValue: PreservePlacementWithSilentEnforcement,
+            storageLocations: new OptionStorageLocation[]{
+                new EditorConfigStorageLocation<CodeStyleOption<AddImportPlacement>>(
+                    "csharp_using_directive_placement",
```
i think this name is good.  but i would use 'using directive' more consistently in the rest of the UI.  This is especially important as C# also has a 'using statement'

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/CodeStyle/CSharpCodeStyleOptions.cs:
```diff
@@ -227,6 +228,19 @@ private static Option<T> CreateOption<T>(OptionGroup group, string name, T defau
+        public static readonly CodeStyleOption<AddImportPlacement> PreservePlacementWithSilentEnforcement =
+            new CodeStyleOption<AddImportPlacement>(AddImportPlacement.Preserve, NotificationOption.Silent);
+
+        public static readonly Option<CodeStyleOption<AddImportPlacement>> PreferredUsingDirectivesPlacement = CreateOption(
```
i don't like that you use Directives here, but Directive below.  can you keep pluralization consistent?

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs:
```diff
@@ -2271,14 +2274,21 @@ private static bool CanReplaceWithReducedName(this NameSyntax name, TypeSyntax r
                 IsNullableTypeInPointerExpression(name, reducedName) ||
                 name.IsNotNullableReplaceable(reducedName) ||
-                IsQualifiedNameInUsingDirective(semanticModel, name, reducedName))
+                (IsQualifiedNameInUsingDirective(semanticModel, name, reducedName) && !IsGlobalAliasQualifiedName(name)))
```
can you extract a named method for this?  to keep all the cases just simple || checks with clear names?

JoeRobich replied:
Moved these changes into https://github.com/dotnet/roslyn/pull/34964

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs:
```diff
@@ -2271,14 +2274,21 @@ private static bool CanReplaceWithReducedName(this NameSyntax name, TypeSyntax r
+            return name is AliasQualifiedNameSyntax aliasName
+                ? aliasName.Alias.Identifier.IsKind(SyntaxKind.GlobalKeyword)
+                : false;
+        }
```
1. shoudl be commented.  
2. can just be `=> name is AliasQual aliasName && ...;`

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DocumentBasedFixAllProvider.cs:
```diff
@@ -14,13 +14,13 @@ namespace Microsoft.CodeAnalysis.CodeStyle
         public override Task<CodeAction> GetFixAsync(FixAllContext fixAllContext)
         {
+            // Disable RS0005 as this is test code and we don't need telemetry for created code action.
+#pragma warning disable RS0005 // Do not use generic CodeAction.Create to create CodeAction
```
can you explain how this is test code?

JoeRobich replied:
I'll remove

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/Editing/VisualBasicImportAdder.vb:
```diff
@@ -25,6 +27,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Editing
         End Function
 
+        Protected Overrides Function GetImportPlacement(options As OptionSet) As AddImportPlacement
+            Return AddImportPlacement.OutsideNamespace
```
maybe a simple omment that VB doesn't support imports inside namespaces.

CyrusNajmabadi commented:
i haven't looked through this all, but this PR is large enough that i think it would benefit from a few initial PRs sent in that break things out to make the final PR much easier to read.  For example, moving to a simple model of having AddImportOptions would make it easy to then add the Placement option and update only the couple of places that need it. 

Similarly, moving the FixAll stuff around looks ike it can be done independently. 

This PR is currently at the point that it triggerred my "too big to fit in my head' point, especially with keeping the big picutre, and the individual logic around at the same time.

JoeRobich commented:
@CyrusNajmabadi Thanks for the review. I will look at extracting a some of these changes as you suggested.

JoeRobich commented on src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.cs:
```diff
@@ -0,0 +1,405 @@
+            }
+
+            var newTrivia = splitIntoLines(firstMemberTrivia)
+                .SkipWhile(trivia => trivia.All(t => t.IsWhitespaceOrEndOfLine()) && trivia.Last().IsKind(SyntaxKind.EndOfLineTrivia))
```
I don't need to remove leading whitespace trivia if it is on the same line.

JoeRobich commented:
Closing and will take these changes through https://github.com/dotnet/roslyn/pull/35009
